crates/m3/src/gadgets/mul.rs

```diff
@@ -0,0 +1,408 @@
+		assert_eq!(BIT_LENGTH, 1 << FP::TOWER_LEVEL);
+		assert_eq!(BIT_LENGTH, UX::BIT_LENGTH);
+		// These are currently the only bit lengths I've tested
+		assert!(BIT_LENGTH == 32 || BIT_LENGTH == 64);
+
+		let x_in = table.add_computed("x_in", pack_fp(x_in_bits));
+		let y_in = table.add_computed("y_in", pack_fp(y_in_bits));
+
+		let generator = UX::generator().into();
+		let generator_pow_bit_len = UX::shifted_generator().into();
+
+		let g_pow_x = table.add_static_exp::<FExpBase>("g^x", &x_in_bits, generator);
+		let g_pow_xy = table.add_dynamic_exp::<FExpBase>("(g^x)^y", &y_in_bits, g_pow_x);
+
+		let out_high_bits = table.add_committed_multiple("out_high");
+		let out_low_bits = table.add_committed_multiple("out_low");
+
+		let out_high = table.add_computed("out_high", pack_fp(out_high_bits));
+		let out_low = table.add_computed("out_low", pack_fp(out_low_bits));
+
+		let g_pow_out_low: Col<FExpBase> =
+			table.add_static_exp("g^(out_low)", &out_low_bits, generator);
+		let g_pow_out_high: Col<FExpBase> = table.add_static_exp(
+			"(g^(2^BIT_LENGTH))^(out_high)",
+			&out_high_bits,
+			generator_pow_bit_len,
+		);
+
+		table.assert_zero("order_non_wrapping", x_in_bits[0] * y_in_bits[0] - out_low_bits[0]);
+		table.assert_zero("exponentiation_equality", g_pow_xy - g_pow_out_low * g_pow_out_high);
+
+		Self {
+			x_in_bits,
+			y_in_bits,
+			out_high_bits,
+			out_low_bits,
+			x_in,
+			y_in,
+			out_high,
+			out_low,
+			_marker: PhantomData,
+		}
+	}
+
+	#[inline]
+	fn populate_internal<P>(
+		&self,
+		index: &mut TableWitnessSegment<P>,
+		x_vals: impl IntoIterator<Item = FP>,
+		y_vals: impl IntoIterator<Item = FP>,
+		fill_input_bits: bool,
+	) -> Result<()>
+	where
+		P: PackedField<Scalar = B128> + PackedExtension<B1> + PackedExtension<FP>,
+	{
+		let mut x_in_bits = array_util::try_map(self.x_in_bits, |bit| index.get_mut(bit))?;
+		let mut y_in_bits = array_util::try_map(self.y_in_bits, |bit| index.get_mut(bit))?;
+		let mut out_low_bits = array_util::try_map(self.out_low_bits, |bit| index.get_mut(bit))?;
+		let mut out_high_bits = array_util::try_map(self.out_high_bits, |bit| index.get_mut(bit))?;
+
+		let mut x_in = index.get_mut(self.x_in)?;
+		let mut y_in = index.get_mut(self.y_in)?;
+		let mut out_low = index.get_mut(self.out_low)?;
+		let mut out_high = index.get_mut(self.out_high)?;
+
+		for (i, (x, y)) in x_vals.into_iter().zip(y_vals.into_iter()).enumerate() {
+			let (res_high, res_low) = UX::mul(x, y);
+			set_packed_slice(&mut x_in, i, x);
+			set_packed_slice(&mut y_in, i, y);
+			set_packed_slice(&mut out_low, i, res_low);
+			set_packed_slice(&mut out_high, i, res_high);
+
+			for bit_idx in 0..BIT_LENGTH {
+				if fill_input_bits {
+					if UX::is_bit_set_at(x, bit_idx) {
+						set_packed_slice(&mut x_in_bits[bit_idx], i, B1::ONE);
+					}
+					if UX::is_bit_set_at(y, bit_idx) {
+						set_packed_slice(&mut y_in_bits[bit_idx], i, B1::ONE);
+					}
+				}
+				if UX::is_bit_set_at(res_low, bit_idx) {
```

Comment:
Instead of the `if`s , just do `set_packed_slice(&mut out_low_bits[bit_idx], i, B1::from(UX::is_bit_set_at(res_low, bit_idx))`